### ISSUE TITLE
Bug in function `all`

### ISSUE THREAD

**Issue by zdlspace0528** [JuliaLang/julia#45562]
I find a bug about `all`.

here is the result.

```
julia> all([3, 3, 3], dims = 1)
1-element Vector{Bool}:
 1
```
The result maybe wrong.

```
julia> for i in 1:10
           @show all(i * ones(Int, 3), dims = 1)
       end
all(i * ones(Int, 3), dims = 1) = Bool[1]
all(i * ones(Int, 3), dims = 1) = Bool[0]
all(i * ones(Int, 3), dims = 1) = Bool[1]
all(i * ones(Int, 3), dims = 1) = Bool[0]
all(i * ones(Int, 3), dims = 1) = Bool[1]
all(i * ones(Int, 3), dims = 1) = Bool[0]
all(i * ones(Int, 3), dims = 1) = Bool[1]
all(i * ones(Int, 3), dims = 1) = Bool[0]
all(i * ones(Int, 3), dims = 1) = Bool[1]
all(i * ones(Int, 3), dims = 1) = Bool[0]
```

**Comment by zdlspace0528** [comment]
We talk about this problem here.[](https://discourse.julialang.org/t/what-does-this-result-mean/82137)

**Comment by LilithHafner** [comment]
The error we want is `ERROR: TypeError: non-boolean (Int64) used in boolean context`

**Comment by LilithHafner** [comment]
From @pfitzseb in the linked [discussion](https://discourse.julialang.org/t/what-does-this-result-mean/82137/6?u=lilith)

>This boils down to
>```
>julia> mapreduce(identity, &, [3,3,3]; dims=1)
>1-element Vector{Bool}:
> 1
>
>julia> mapreduce(identity, &, [2,2,2]; dims=1)
>1-element Vector{Bool}:
> 0
>```
>and definitely is a bug.

The problem is that we are using `&` and `|` for all and any when we should be suing `&&` and `||`.

**Comment by LilithHafner** [comment]
We should be using something that errors on non-booleans. We can write up custom functions and use these instead:
```julia
boolean_and(a::Bool, b::Bool) = a & b
boolean_and(a::Bool, b) = throw(TypeError(:boolean_and, Bool, b))
boolean_and(a, b) = throw(TypeError(:boolean_and, Bool, a))
boolean_or(a::Bool, b::Bool) = a | b
boolean_or(a::Bool, b) = throw(TypeError(:boolean_or, Bool, b))
boolean_or(a, b) = throw(TypeError(:boolean_or, Bool, a))
```

**Comment by GunnarFarneback** [comment]
Notice that the investigation revealed two separate bugs with reductions over an axis.

1. This should error like its non-axis version:
```
julia> all([3, 3, 3], dims = 1)
1-element Vector{Bool}:
 1

julia> all([3, 3, 3])
ERROR: TypeError: non-boolean (Int64) used in boolean context
```

2. The downstream `mapreduce` also gives an incorrect result along an axis:
```
julia> mapreduce(identity, &, [3,3,3], dims = 1)
1-element Vector{Bool}:
 1

julia> mapreduce(identity, &, [3,3,3])
3
```

**Comment by pfitzseb** [comment]
The `reducedim` part of this bug is caused by
https://github.com/JuliaLang/julia/blob/bd8dbc388c7b89f68838ca554ed7ba91740cce75/base/reducedim.jl#L206-L207
which causes the returned array to be unconditionally initialized as a `Array{Bool}`, even though `&`/`|` aren't guaranteed to return booleans.

**Comment by LilithHafner** [comment]
@GunnarFarneback, the problem you labeled 2 is at the intersection of this issue and #45566.
The linked PR fixes 1 and turns 2 from a wrong answer into an error.

**Comment by GunnarFarneback** [comment]
I assume you mean the opposite. It would be good if it added tests for both problems.

**Comment by GunnarFarneback** [comment]
Thanks, I misunderstood. Turning a wrong answer into an error sounds a bit scary, even if both are incorrect.

**Comment by StefanKarpinski** [comment]
Unless I'm missing something, the middle method of each of these is unnecessary. You can just do:
```julia
boolean_and(a::Bool, b::Bool) = a & b
boolean_or(a::Bool, b::Bool) = a | b
boolean_and(a, b) = throw(TypeError(:boolean_and, Bool, a))
boolean_or(a, b) = throw(TypeError(:boolean_or, Bool, a))
```
Could generalize this to
```jl
function boolean_op(f::Function)
    op(a, b) = throw(TypeError(nameof(f), Bool, a))
    op(a::Bool, b::Bool) = f(a, b)
end
```
Although that does produce annoyingly ugly stack traces on account of the inner function.

**Comment by pfitzseb** [comment]
This is unrelated to correctness, but shouldn't `all`/`any` short-circuit? There really is no need to check the whole array after all (although you'll of course end up with a sort of ragged iteration scheme which could be slower in some cases, I guess).

**Comment by LilithHafner** [comment]
> Unless I'm missing something, the middle method of each of these is unnecessary. You can just do:

I don't like this error:
```
julia> boolean_and(true, 5)
ERROR: TypeError: non-boolean (Bool) used in boolean context
```
> Could generalize this

Yes, though I do err on the side of cleaner error messages.

> shouldn't all/any short-circuit?

Thanks @pfitzseb! I expect tat this would better and obviate the whole issue. I'll look into it more.

**Comment by raminammour** [comment]
> This is unrelated to correctness, but shouldn't `all`/`any` short-circuit? There really is no need to check the whole array after all (although you'll of course end up with a sort of ragged iteration scheme which could be slower in some cases, I guess).

I agree with the short-circuting behavior being more efficient. But then the check on an array for example needs to be done separately, depending on how the short-circuiting is achieved. `false && :hello` would return false right away. So will need to check that an array is an `Array{Bool}` before, otherwise bugs could be hidden again...

**Comment by LilithHafner** [comment]
We already have `all([false, :hello]) == false`. I think that this is okay because it is explicitly documented to be short circuited. It can also sometimes be somewhat challenging to determine whether an input is actually all Bools (e.g.  `all(function, itr)` may require executing the function on every element of itr).

**Comment by LilithHafner** [comment]
I think that it is appropriate for `all(f, [a,b,c,d])` to be the same as `f(a) && f(b) && f(c) && f(d)`.

**Comment by StefanKarpinski** [comment]
Bump, would be good to get a fix in for this.

**Comment by MilesCranmer** [comment]
This can be closed. (I don't have triage privileges)

```julia
julia> all([3, 3, 3]; dims=1)
ERROR: TypeError: non-boolean (Int64) used in boolean context
Stacktrace:
  [1] and_all
    @ ./reduce.jl:32 [inlined]
```

**Comment by adienes** [comment]
I think to fix
```
julia> mapreduce(identity, |, [1, 1, 1]; dims=1)
1-element Vector{Bool}:
 1
```
we still need #58418

**Comment by MilesCranmer** [comment]
Different issue, I think? `all`/`any` no longer go through that particular `mapreduce` path, so it solves the original issue, even if it didn't solve it by fixing mapreduce on `|`

```julia
ERROR: TypeError: non-boolean (Int64) used in boolean context
Stacktrace:
  [1] and_all
    @ ./reduce.jl:32 [inlined]
  [2] macro expansion
    @ ./reducedim.jl:281 [inlined]
  [3] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [4] _mapreducedim!(f::typeof(identity), op::typeof(Base.and_all), R::Vector{Bool}, A::Vector{Int64})
    @ Base ./reducedim.jl:280
  [5] mapreducedim!
    @ ./reducedim.jl:297 [inlined]
  [6] _mapreduce_dim
    @ ./reducedim.jl:344 [inlined]
  [7] mapreduce
    @ ./reducedim.jl:330 [inlined]
  [8] _all
    @ ./reducedim.jl:1012 [inlined]
  [9] _all
    @ ./reducedim.jl:1011 [inlined]
 [10] kwcall(::@NamedTuple{dims::Int64}, ::typeof(all), a::Vector{Int64})
    @ Base ./reducedim.jl:995
 [11] top-level scope
    @ REPL[1]:1
```

i.e., I think this issue by itself could be closed.
